packages/components/src/navigation/index.js

```diff
@@ -30,18 +30,42 @@ const Navigation = ( { activeItemId, children, data, rootTitle } ) => {
 	};
 
 	const mapItems = ( itemData ) => {
+		const groupings = [];
 		const items = new Map(
-			[
-				{ id: 'root', parent: null, title: rootTitle },
-				...itemData,
-			].map( ( item ) => [ item.id, appendItemData( item ) ] )
+			[ { id: 'root', parent: null, title: rootTitle }, ...itemData ]
+				.filter( ( item ) => {
+					if ( item.type === 'grouping' ) {
+						item.children = [];
+						groupings.push( item );
+						return false;
+					}
+					return true;
+				} )
+				.map( ( item ) => [ item.id, appendItemData( item ) ] )
 		);
 
 		items.forEach( ( item ) => {
 			const parentItem = items.get( item.parent );
 			if ( parentItem ) {
-				parentItem.children.push( item );
-				parentItem.hasChildren = true;
+				if ( item.group ) {
+					const grouping = groupings.find(
+						( group ) => group.id === item.group
+					);
+					if ( grouping ) {
+						grouping.children.push( item );
+					}
+				} else {
+					parentItem.children.push( item );
+					parentItem.hasChildren = true;
+				}
+			}
+		} );
+
+		groupings.forEach( ( grouping ) => {
+			const parentItem = items.get( grouping.parent );
+			if ( parentItem ) {
+				parentItem.groupings = parentItem.groupings || [];
+				parentItem.groupings.push( grouping );
```

Comment:
This is will just become more complex over time, ideally we should move a way from "config" based API and use a component composition based one.

We have the same issue with DropdownMenu component where it's config based like the navigation one and unfortunately, we hit the limit on several occasions everytime we have a slightly different use-case.

Reply:
@youknowriad I've seen that `DropdownMenu` has 2 different interfaces, with config or children.
Building upon the children one, we could do something like this for the Navigation:
```jsx
<Navigation>
  { ( { navigateTo, currentLevel, activeItem } ) => ( <Fragment>
    <NavigationLevel
      level="root"
      title={ __( 'Home' ) }
      isCurrentLevel={ 'root' === currentLevel }
    >
      <NavigationItem
        slug="item-1"
        title={ __( 'Item 1' ) }
        badge={ 0 }
        onClick={ navigateTo( 'sub-menu' ) }
        isActive={ 'item-1' === activeItem }
      />
      <NavigationItem
        slug="external-link"
        title={ __( 'External Link' ) }
        href="#"
        isActive={ 'external-link' === activeItem }
      />
    </NavigationLevel>

    <NavigationLevel
      level="sub-menu"
      title={ __( 'Sub Menu' ) }
      isCurrentLevel={ 'sub-menu' === currentLevel }
    >
      <NavigationBackButton onClick={ navigateTo( 'root' ) } />
      <NavigationItem
        slug="child-1"
        title={ __( 'Child 1' ) }
        isActive={ 'child-1' === activeItem }
      />
    </NavigationLevel>
  </Fragment> ) }
</Navigation>
```

With the `navigateTo` function controlling the whole toggle state and animation when traversing the menu hierarchy, and the various `NavigationItem` that could be either simple buttons or render anything inside.

Reply:
Yes something like that.

I actually think the current config implementation could be built on top of a more flexible API based on components composition. So we'd have one API and potentially a shortcut (not sure if needed on Core though).

I think @ItsJonQ had some similar ideas on explorations maybe on his G2 project?

Reply:
If we go this route though, it's probably a refactor that should happen separately and not block the current PR.

Reply:
@youknowriad Do you think such refactor can happen _after_ merging the [main Navigation PR](https://github.com/WordPress/gutenberg/pull/24920)?

I know I'm kinda pushy with shipping it, but it's experimental and not used anywhere, and I believe it would be much easier to develop and iterate in master, as opposed to feature branches that would eventually become too unwieldy to review. 🙂

Reply:
It seems to me that if we all agree that the refactor is necessary it's better to wait for it to happen, especially because it's a completely different API and all users will have to change their code as a consequence.

Reply:
>Do you think such refactor can happen after merging the main Navigation PR?

Since the current implementation [has reached some stability](https://github.com/WordPress/gutenberg/issues?q=is%3Aopen+is%3Aissue+label%3A%22%5BFeature%5D+Navigation+Component%22), a large refactor won't cause conflicts with ongoing work. We've been using `npm link` to the `feature/navigation` branch, so I'm ok to do a refactor before the merge.

>we could do something like this for the Navigation:

The challenge that jumps out to me is that not every navigation view will have sub menus and some will have multiple submenus. How would you compose a render function to handle all those possibilities? I will have a ponder on that today and take inspiration from @ItsJonQ 's code to see if I can find a solution.

Thanks for the feedback @Copons @youknowriad !

Reply:
I've opened a (very rough!) draft of the approach I outlined earlier: https://github.com/WordPress/gutenberg/pull/25057

Apologize for its state, but I wanted to push out something to avoid losing a day (a whole weekend even?) due to timezone differences. 🙂

Reply:
Thanks for that draft @Copons!  Overall I do like the compositional approach more.

However, we do shift most of the responsibility for determining active items and active levels to the consumer.  If we're okay with this then I'm all for the refactor.

Reply:
> However, we do shift most of the responsibility for determining active items and active levels to the consumer. If we're okay with this then I'm all for the refactor.

is this something that we can provide using hooks and context. an API like that could be awesome:

```
const [ screenId, params ] = useActiveScreen();
```

Reply:
@youknowriad the current version of the composition approach #25057 does indeed use the Context API.
I haven't moved the "active" logic into a custom hook as it's very simple and straightforward, but nothing blocks us to do so.

At the moment, the consumer doesn't need to know or do anything about the internal state of the navigation (although they optionally can).